docs/architecture.md

```diff
@@ -0,0 +1,215 @@
+# System Architecture
+
+## Core Philosophy
+
+### Idiomatic Go
+- **Simplicity and Clarity**: Straightforward code over clever abstractions
+- **Explicit Dependencies**: Function signatures and struct fields make dependencies obvious
+- **Error Handling**: Errors include context and are wrapped for debugging
+- **Pure Resource Builders**: Resource functions are deterministic - same input produces same output
+- **Structured Concurrency**: Use sync.WaitGroup for coordinated goroutines with clear lifecycle
+
+### Kubernetes Operator Safety
+- **Finalizers for Lifecycle**: Prevent resource deletion until cleanup is complete
+- **Owner References for Cleanup**: Automatic garbage collection of child resources when parent is deleted
+- **Idempotent Reconciliation**: Safe to run reconcile loop multiple times - converges to desired state
+- **Status Subresource**: Observed state lives in status, never in spec
+- **Infrastructure-Only Concerns**: Operator manages compute resources, not application logic or startup dependencies
+
+## Design Principles
+
+### Clean Separation of Concerns
+- **Operator**: Manages Kubernetes resources (Deployments, StatefulSets, Services, HPAs)
+- **Multigres Components**: Handle their own application logic, readiness, and inter-service dependencies
+- **No Application Knowledge**: Operator doesn't orchestrate Multigres component startup order
+- **Eventually Consistent**: All resources created concurrently; components become ready when dependencies are available
+
+### Layered Architecture
+- **Controller Layer**: Orchestrates reconciliation, manages finalizers and status updates
+- **Reconciler Layer**: Component-specific reconciliation (etcd, multigateway, multiorch, multipooler)
+- **Resource Builder Layer**: Pure functions that construct Kubernetes manifests
+- **Parallel Reconciliation**: All component reconcilers run concurrently via goroutines
+
+### Testability
+- **Pure Functions**: Resource builders are deterministic and table-test friendly
+- **Test Helpers**: Mocks and test doubles enable testing without external dependencies
+- **Integration Tests**: envtest provides real Kubernetes API for controller testing
+- **Minimal Interfaces**: Each reconciler has consistent, simple signature
+
+## System Architecture
```

Comment:
a bunch of this section appears to be duplicated, perhaps the README one can be just high level? WDYT?

Reply:
I agree, let me keep this one as is, and update the one in README to be more high level 👍

Reply:
This is done through multiple commits, but essentially:
https://github.com/numtide/multigres-operator/pull/1/commits/1956fc8980ef4796c1a48f88690702f474061299